### ISSUE TITLE
Improve selection welcome page

### ISSUE THREAD

**Issue by kelson42** [openzim/mwoffliner#449]
Using `mwoffliner` with selection of articles, ie. with command line argument `--articleList` has currently following consequences:
* Welcome page is a simple list of the articles (in the ZIM)
* Fill the ZIM only with the articles of the list

This is good, but the welcome page is really a problem because:
* it is "ugly" and not really helpful
* it does not scale.

I propose to following things: Replace the list of title by a full page of tilled pictures. Each picture would represent an article and in fact point to it. Because we can not create a page with 1 millions images, we would display only the first one, and scrolling down will extend the page with more pictures. All of this would be sorted from the first article in the list given to the last one.

**Comment by kelson42** [comment]
Technical details:
* Most of the images should be anyway in the articles anyway. Probably the size increase should be limited.
* We could anyway limit the capacity of the welcome page to maybe something like 1000 images.
* We could use framework Masonery https://masonry.desandro.com/
* We can identify an image for most of the article using the https://www.wikidata.org/wiki/Property:P18 of Wikidata. Simple SPARQL requests should allow to do that.

**Comment by Popolechien** [comment]
Sounds good. I have a couple of top-line comments:
- In some cases the Image Property can return multiple items (e.g. [https://www.wikidata.org/wiki/Q1071](https://www.wikidata.org/wiki/Q1071));
- We should maje sure that this tiled menu is also generated with the "nopic" selection (ie no image in the file except for this one landing page);
- There need be some text overlay on each image;
- I think 100 images is more than enough - for users on a phone, that's already of lot of scrolling;

**Comment by kelson42** [comment]
@Popolechien I agree, and the images should be loaded only if displayed (so dynamically). What would you write in this text overlay? When should it appear?
@ISNIT0 Does that sounds doable?

**Comment by Popolechien** [comment]
Well the text overlay should simply describe exactly what the image links to, e.g. Michael Jackson or Geography. Then there is also the problem of font color vs. background. In the below patchwork (I took a random of some of the most viewed, images are properties of the corresponding Wikidata items) I had to move the text around on the China image so as not to have white on white. Any dynamic generation would have to deal with it, wouldn't it?

![patch](https://user-images.githubusercontent.com/21675885/49065644-e6cdfc80-f21e-11e8-85e8-fcf523bd76aa.jpg)

**Comment by ISNIT0** [comment]
We could solve the text issue with a translucent overlay behind the text.
The dynamic loading would be impossible to do in noJS environments, but fine the rest of the time.

**Comment by Popolechien** [comment]
Ok ELI5 please: what noJS environments are there out there (ie how many users would we frustrate, and who would they be)?

**Comment by ISNIT0** [comment]
@mossroy @jaifroid

I think we need to count browser extensions as noJS?

**Comment by ISNIT0** [comment]
@kelson42 
- Will there ever be a case where there are no images?
- If the article list has 200 items, do we only show 100? Which 100?
- Do we want to have search functionality? Plaintext search of titles, categories?
- We _could_ make the articleList more than just a list, it could have sections and section titles (similar to markdown) or category separators to make the landing page better?

**Comment by Jaifroid** [comment]
@ISNIT0 @Popolechien @mossroy

Kiwix JS is currently unable to display active content generated by JavaScript in the ZIM. Most extensions ban inline JavaScript (used extensively in the UIs of TEDx, Gutenberg and other ZIMs). We have support for JS in clients that support Service Workers, but they will never work in extensions given how these proprietary UIs are currently coded.

Therefore the move towards this kind of JS-generated UI is progressively locking us out of accessing the content from the ZIMs. For example, recent Gutenberg ZIMs are pretty inaccessible to Kiwix JS clients of any flavour as explained in https://github.com/openzim/gutenberg/issues/75 . And it is very hard to access the content of TEDx ZIMs (we can now play the videos, as they don't depend on JavaScript).

I have coded a workaround in Kiwix JS Windows, by throwing up an old-style traditional alphabetical index -- you can see screenshots in https://github.com/kiwix/kiwix-js-windows/issues/64.

So, what we'd ideally like is a `noscript` version of pages that have active content, a minimally accessible version that looks reasonable and doesn't present the user with a "broken" interface. We can detect the noscript version and put it up when the client can't run JS-in-the-ZIM.

**Comment by ISNIT0** [comment]
@Popolechien Is this related? https://github.com/openzim/mwoffliner/issues/405

**Comment by Popolechien** [comment]
Yes. 

Le 27 nov. 2018 à 13:50, à 13:50, Joe Reeve <notifications@github.com> a écrit:

**Comment by Popolechien** [comment]
@ISNIT0 Yes. Different side of the frontpage issue though.

**Comment by kelson42** [comment]
@Jaifroid @mossroy I do not understand the problem with kiwix-js and javascript in pages. We obviously need to be able to run javascript available inside the HTML content delivered by ZIM files. What is so different with the Gutenberg ZIM file?

**Comment by Jaifroid** [comment]
> We obviously need to be able to run javascript available inside the HTML content delivered by ZIM files.

@kelson42 We have managed to get by without doing this since the beginning of Kiwix HTML5 / JS! It is not *needed* to access the content of Wikipedia / Wikivoyage / WikiMed. The only bit of JS in those files is one to open and close the heading sections. It was causing problems because it used inline script, banned in extensions. @ISNIT0 helped us to fix that, and support clients that can't run the script (with a `noscript`) section.

Only recently (in the last couple of weeks) have we enabled a version that can execute JS inside the ZIM, but it only works in Service Worker. SW can only be accessed if you serve Kiwix JS from localhost or a proper web server with https://, not from file:// protocol, and it's not available in extensions either.

I have an experimental branch of Kiwix JS without Service Worker (jQuery mode) that tries to turn inline scripts into attached scripts and insert them into the DOM, so that we avoid the problem with inline scripts. It manages to execute PhET experiments, and the open-close heading code in Wikimedia ZIMs, but fails to run the UI for PhET, Gutenberg, TEDx and dirtybiology amongst others.

The Gutenberg problem is that if we can't run the UI, then there are no author pages, because these are constructed dynamically by the UI. That is why I requested the author's surname(s) to be added to the title of Gutenberg "book" pages. We can *search* the dirEntry titles, but this is only helpful (for books) if the dirEntries are sorted alphabetically by author's surname(s).

I tried hard to "convert" those UIs, but they contain code that is very hard to "patch". Without Service Worker, we can't intercept the calls that the DOM is making, except by patching the function that makes the call.

For example, say the UI needs "../JS/myProprietaryUI.js", and requests that dynamically, in JavaScript, using an XHR / Fetch request (this is the case in one of the UIs), then it will fail unless I can intercept that request, extract the needed piece of JavaScript, and insert it into the function that called it. In Service Worker this "just works", because SW has an intercept API, but there is no equivalent intercept if SW is not available.

So, we either do a lot of patching work, altering JS files in the ZIM as we extract them (hard, non-generic!), or we provide an alternative `noscript` interface, either in the client itself (https://github.com/kiwix/kiwix-js-windows/issues/64), or in the ZIM, or preferably both, by providing a minimal working UI that doesn't require that the client execute JS and a working, useful sort order for the dirEntries in all ZIMs.

**Comment by kelson42** [comment]
> @kelson42
>     * Will there ever be a case where there are no images?

No we want this, all the time like @Popolechien said earlier.
 
>     * If the article list has 200 items, do we only show 100? Which 100?

The first 100 (having a picture provided by Wikidata) of the list of article titles.

>     * Do we want to have search functionality? Plaintext search of titles, categories?

This is not necessary.

>     * We _could_ make the articleList more than just a list, it could have sections and section titles (similar to markdown) or category separators to make the landing page better?

Please open a separate ticket to that. This ticket is quit urgent and I do not want to slow down too much the implementation with long discussion about the requirements. That said, this might be good ideas... but this is not clear to me (1) how this would looks like (2) on which data you would base this categorisation (3) how this would concretly work.

**Comment by mossroy** [comment]
I'll try to give an overview on javascript support (regarding kiwix-js, but not only).

We totally agree that using javascript can improve the user experience, but there must be some fallbacks.
We do what we can to support javascript in kiwix-js, but there are restrictions that are not in our hands.

TLDR : use all the javascript you need to improve the UI, but please : 

- **do not use inline javascript** (including inline event handlers) : all the javascript code should be in javascript files, and event handlers should be attached in these files (not by using `<button onclick="function()">`)
- ideally, design the UI so that it **works in fallback mode if these javascript files are not loaded/executed** (with less fancy features, of course)
- if it's really necessary, use `<noscript>` tags to do so

Here are a few technical reasons : 

1. Old browsers without ServiceWorker support
Javascript is 100% supported in ServiceWorker mode of kiwix-js. But this mode is not supported by some old browsers/platforms that can't be upgraded : **Windows 10 Mobile** for sure (because it uses an old version of Edge), probably **Ubuntu Touch** (old versions of Chromium engine), Firefox OS (OK, it does not count any more), and old versions of any browser.
For these platforms, we need to use another mode (called jQuery mode), where we need to parse the DOM (with jQuery), try to find all the needed assets the page needs (images, CSS etc), read them in the ZIM file, and manually inject them in the DOM. It might sound crazy because we need to code what a browser naturally does, but we don't have another choice.
We obviously can't find 100% of the assets, and are not able to do that very efficiently. It would be particularly complicated to handle javascript, which is why it is currently ignored. @Jaifroid does hard work to handle some ZIM files, but it's a tremendous work for a support that will certainly be incomplete.
But at least we should be able to handle `<noscript>` tags in jQuery mode.

2. Restrictions on ServiceWorker support
There are some cases where the browser engine supports the ServiceWorker technology, but does not enable it : if kiwix-js is used through file:// or http:// (not https) protocols, and (most important) in **Firefox extensions** (see https://bugzilla.mozilla.org/show_bug.cgi?id=1344561)

3. Content Security Policies (CSP)
**Browser extensions** enforce restrictions on javascript execution, through CSP, that disallow inline javascript (including inline event handlers).
They do that for security reasons, and it's not the only use-case where CSP can be applied : it was the case for Firefox OS applications too.
Some CSP might be applied in other contexts in the future.

4. (outside the scope of kiwix-js) Browsers that do not handle javascript
If we use kiwix-serve for example, javascript execution can be blocked by the browser, even if it supports it : because you use Tails browser or a browser extension that does that, or because you have an old computer (where it would be too slow) and/or an old browser (that would not handle it properly).
In some of these cases, `<noscript>` tags can be a workaround, but not all of them

To conclude, if we don't want to drop support for these devices/platforms/usages, we need to design the ZIM contents so that they make use of all the power of javascript when it is available, but have a fallback mode when it is not.
That's what @ISNIT0 recently did for the section toggling feature (see https://github.com/openzim/mwoffliner/issues/324)
mwoffliner certainly generates the most used ZIM files, and wikimedia contents are already mostly "nojs-compliant" : it would be a pity to loose that capability.

**Comment by ISNIT0** [comment]
By no means complete, but this is where we're at so far:
https://framadrop.org/r/7tIckfEx0j#IR62qiivR/bG/qT9Q8HDPSqdPYrS1alcsaN9VTW5rc0=

Known issues:
- Articles are not displayed in the same order as in the `articleList` but are sorted alphabetically
- No sensible handling for non wikipedia sites yet

**Comment by kelson42** [comment]
@ISNIT0 @tamasfabi, Thank you for this promising first version. Here are my remarks (beside the issue you have already mentioned):
* No list at the end
* Background should be invisible, pictures should stick nicely to each other to the full page
* No title at the top of the page

**Comment by ISNIT0** [comment]
Here's a ZIM of the above mentioned issues solved:
https://framadrop.org/r/TwJI5CzYjk#g/zoZzfJSu57jX5zNLX54eprwMLOy2tw+tFyHRmjnRo=

**Comment by kelson42** [comment]
@ISNIT0 Have you tried it with kiwix-desktop? To me it display only one column with the pictures (extremely magnified) in a (vertical) row, not a patchwork display at all.

Otherwise you could also try it now with a real case of article selection: https://download.kiwix.org/wp1/enwiki_2018-12/tops/1000

**Comment by Popolechien** [comment]
Ok so I just tested it on Kiwix-desktop macOS, Kiwix-JS and android:
- on macOS, only one column, does not re-organize but zooms as I expand window;
- on Kiwix-JS I have a nice tile layout (what we expect);
- on android a single column as well that fits the window size.

It's actually a good thing we use @ISNIT0 's selection as the first article is that of a Manuel Torres, and seeing his face first thing (and taking up all screenspace) is a bit creepy (no offense meant to the gentleman). I would suggest we make the minimum column size to be two images (that's sort of what they have on the WP app).

- Another good learning from the WP app is that they include a few words of introduction (I suspect taken directly taken from the article itself). Since we take the image from Wikidata, couldn't we also  add the item descriptor (usually 2-4 words, in the above case "Puerto Rican politician") in slightly smaller font (or not bold vs. bold) next to the name?

**Comment by kelson42** [comment]
1 - We will see, this is really an other topic, this list is not created anyway by `mwoffliner`
2 - All sub-selections ZIM (not clear to me if we will use it for `nopic`)
3 - This is the role of the reader, not of the ZIM file IMO.

**Comment by ISNIT0** [comment]
@kelson42 Here is the latest version:

What should we do when no images are found? Show an old-style welcome page?

**Comment by kelson42** [comment]
@ISNIT0 Skip the articles with no images and go ahead down in the list. If at the end, you have less than THRESHOLD images available, fallback to the text mode.

**Comment by ISNIT0** [comment]
Outputs of #469:

With ~100 articles: 
https://framadrop.org/r/RvjnhZNL-Z#02myCmB4E16WWumDvjcS/X2S8Q81RdPIKK9Ovu/na64=

With <10 articles: (switch to link only mode)
https://framadrop.org/r/l94uKjTa15#I9u7zyQUqeZh90uATxvVF90KPonrQhqS8oYRT0SQKUc=

**Comment by Popolechien** [comment]
Looks good on macOS and Kiwix-JS: is there a limit to the number of images by row (4, by the look of it)? On a large screen I can get 2, 3 or 4 images per row, not more (and not less). Not really a blocker, but just a remark.

**Comment by kelson42** [comment]
I re-open the issue because I see a few things which are not working well:
* The background seems to be white instead of black (if pictures, default nopic version should still be white)
* The effect of zoom by going over a tile has disappear, this was great (an important IMO)
* In Kiwix Desktop, tiles stick to each other but if I open one an come back, then a padding of a few pixels appear beetween all tiles.

**Comment by ISNIT0** [comment]
@kelson42 
> In Kiwix Desktop, tiles stick to each other but if I open one an come back, then a padding of a few pixels appear beetween all tiles.

I can't reproduce this, it may be a webview rendering issue (I don't know what Kiwix Desktop uses for viewing the ZIM files) or related to a specific window size.

**Comment by kelson42** [comment]
With last nighly of Kiwix Desktop on Ubuntu: https://download.kiwix.org/nightly/2018-12-23/

You can see the padding beetween the images. I see the same with kiwix-serve and FF64.0.

![scr](https://user-images.githubusercontent.com/1029718/50397261-6c39c380-076f-11e9-8407-0892d4809777.png)

**Comment by ISNIT0** [comment]
Okay, I am unable to reproduce on OSX 2.0.9 or using kiwix-serve in Chrome or Firefox.

![image](https://user-images.githubusercontent.com/5173131/50404176-e3398100-079c-11e9-8088-2784d099b022.png)

**Comment by ISNIT0** [comment]
@kelson42 
Latest ZIM file (with black background):
https://framadrop.org/r/TMmynJ6L0Y#fHfKYWpf3pWx+5DL5riz2Fz929JwesIyoxmLMAyzo94=

**Comment by Popolechien** [comment]
Works fine by me (macOS, Kiwix-JS and Android). 
@kelson42 why should there be a black background rather than the customary white? It does look kinda weird, particularly if the image has no background.
![capture d ecran 2018-12-26 a 18 50 29](https://user-images.githubusercontent.com/21675885/50453185-27648700-093f-11e9-9940-e5ff7b6ad43b.png)

**Comment by kelson42** [comment]
Works mostly fine now :) Thx.
